src/main/scala/tsiHost/TsiHostWidget.scala

```diff
@@ -0,0 +1,335 @@
+
+  val io: TLTSIHostMMIOFrontendBundle
+
+  val params: TSIHostParams
+
+  // tsi format input/output queues
+  val txQueue = Module(new Queue(UInt(SerialAdapter.SERIAL_TSI_WIDTH.W), params.txQueueEntries)) // where is the queue being dequeued (to the SerialAdapter)
+  val rxQueue = Module(new Queue(UInt(SerialAdapter.SERIAL_TSI_WIDTH.W), params.rxQueueEntries)) // where is the queue being enqueued (from the SerialAdapter)
+
+  // connect queues to the backend
+  io.serial.out <> txQueue.io.deq
+  rxQueue.io.enq <> io.serial.in
+
+  // memory mapped registers and connections to the queues
+  regmap(
+    TSIHostWidgetCtrlRegs.txQueueOffset -> RegFieldGroup("txdata", Some("Transmit data"),
+                                           NonBlockingEnqueue(txQueue.io.enq, 64)),
+    TSIHostWidgetCtrlRegs.rxQueueOffset -> RegFieldGroup("rxdata", Some("Receive data"),
+                                           NonBlockingDequeue(rxQueue.io.deq, 64))
+  )
+}
+
+/**
+ * Top level class that uses a register router to connect MMIO to the core. This instantiates
+ * the implementation module and bundle
+ *
+ * @param beatBytesIn amount of bytes to send per beat
+ * @param params the TSI parameters for the widget
+ */
+class TLTSIHostMMIOFrontend(val beatBytesIn: Int, params: TSIHostParams)(implicit p: Parameters)
+  extends TLRegisterRouter(
+    base = params.mmioBaseAddress,
+    devname = "tsi-host-mmio-frontend",
+    devcompat = Seq("ucbbar,tsi-host"),
+    beatBytes = beatBytesIn)(
+      new TLRegBundle(params, _)    with TLTSIHostMMIOFrontendBundle)(
+      new TLRegModule(params, _, _) with TLTSIHostMMIOFrontendModule)
+
+/**
+ * Module to decouple the SerialAdapter and TLSerdesser from the Frontend
+ * connection type (MMIO, other...)
+ *
+ * @param params the TSI parameters for the widget
+ */
+class TLTSIHostBackend(val params: TSIHostParams)(implicit p: Parameters)
+  extends LazyModule
+{
+  // module to take in a decoupled io tsi stream and convert to a TL stream
+  val serialAdapter = LazyModule(new SerialAdapter)
+  // This converts the TL signals given by the serial adapter into a decoupled stream
+  val serdes = LazyModule(new TLSerdesser(
+        w = params.serialIfWidth,
+        clientPortParams =  params.serdesParams.clientPortParams,
+        managerPortParams = params.serdesParams.managerPortParams,
+        hasCorruptDenied = params.serdesParams.hasCorruptDenied))
+
+  // you are sending the TL request outwards... to the serdes manager... then to a serial stream
+  serdes.managerNode := TLSourceSetter(params.mmioSourceId) := TLBuffer() := serialAdapter.node
+  // send TL transaction to the memory system on this side
+  // create TL node to connect to outer bus
+  val externalClientNode = serdes.clientNode
+
+  lazy val module = new LazyModuleImp(this) {
+    val io = IO(new Bundle {
+      val adapterSerial = new SerialIO(SerialAdapter.SERIAL_TSI_WIDTH)
+      val serdesSerial =  new SerialIO(params.serialIfWidth)
```

Comment:
Ok now I see why the params weren't being used before but I think it means that the `hostparams.serialIfWidth` should be renamed if its really `hostparams.offChipSerialIfWidth`

Reply:
Yup. I renamed this.